### PR TITLE
Add UI to Deployment form with bootstrap.

### LINKED ISSUE #46: Ability to use branch when deploying using UI
Modify the form to look like this:

Let's use bootstrap to improve the look and feel.

### LINKED ISSUE #54: Fix text in dockup_ui deployment form
1. When not using github, the help text still says "Please enter a github project" : 
image
2. When using github , the label "Url:" is missing above github url

### PR DESCRIPTION
**Pull request by Atul9**
Add UI for deployment form.
Add bootstrap.

closes #46 
fixes #54

### CODE REVIEW

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 1]
```diff
@@ -19,10 +21,15 @@ class DeploymentForm extends Component {
   }
 
   displayHelpText() {
-    if (!this.state.username || !this.state.repository) {
-      return "Please enter a github project to deploy";
+    if (!this.state.username || !this.state.repository || !this.state.branch) {
+      return "Please enter a github project and branch to deploy";
+    } else if ((this.props.isGithub == false) && (!this.state.url || !this.state.branch)){
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: can these logic checks be extracted out into simpler methods? It's hard to parse the flow with double negations and multiple levels of conditionals.

Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: That is, instead of:

```javascript
(this.props.isGithub == false) && (!this.state.url || !this.state.branch)
```

they can be changed to:

```javascript
isGithub () {
  // agree that this is a very simple thing to extract out, so feel free to use your judgement to leave this out
  this.props.isGithub
}

isAnyFieldEmpty () {
  !(this.state.url && this.state.branch)
}
```

and so on.

Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Also, if you see yourself using more than one `if...else`, consider using a `switch..case`

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 2]
```diff
@@ -31,14 +38,57 @@ class DeploymentForm extends Component {
...
+  handleUrlChange(event) {
+    this.setState({url: event.target.value});
+  }
+
+  handleBranchChange(event) {
+    this.setState({branch: event.target.value});
+  }
+
   render() {
+    var partial;
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: prefer `let` or `const`

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 3]
```diff
@@ -31,14 +38,57 @@ class DeploymentForm extends Component {
...
+    if(this.props.isGithub == true) {
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Aim for the least number of `if` conditions in your code. If you see that you had to use a conditional in multiple places, it's a good chance to consider splitting the component. Here, you can have two components instead of one. Google "composing react components".

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 4]
```diff
@@ -31,14 +38,57 @@ class DeploymentForm extends Component {
...
+      partial = <div className="col-md-12">
+                  <table>
+                    <tbody>
+                      <tr>
+                        <td>
+                          <b> https://github.com/ </b>
+                        </td>
+                        <td>
+                          <input name="username" onChange={this.handleUserNameChange.bind(this)} className="form-control"/>
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: prefer putting `bind` calls inside constructor

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 5]
```diff
@@ -31,14 +38,57 @@ class DeploymentForm extends Component {
...
-        <input name="username" onChange={this.handleUserNameChange.bind(this)}/>/
-        <input name="repository" onChange={this.handleRepositoryChange.bind(this)} />
-        <button name="deploy" onClick={this.handleClick.bind(this)} disabled={(!this.state.username || !this.state.repository)}>Deploy</button>
+      <div>
+        {partial}
+        <br />
+        <b>
+          Branch :
+        </b>
+        <br />
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: avoid using `<br/>` tags as much as possible. prefer `div`s or any of the block HTML tags (`p`, `div` etc)

File: apps/dockup_ui/web/static/js/components/deployment_index.js [hunk 6]
```diff
@@ -14,11 +16,28 @@ class DeploymentIndex extends Component {
     deploymentsArray = deploymentsArray.concat(newDeployment)
     this.setState({deployments: deploymentsArray});
   }
+
+  handleClick(event) {
+    if(this.state.isGithub == true) {
+      this.setState({isGithub: false})
+    } else {
+      this.setState({isGithub: true});
+    }
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_index.js]: Can this be simplified to `this.setState({isGithub: !this.state.isGithub})`? Will that work?

File: apps/dockup_ui/web/static/js/components/deployment_index.js [hunk 7]
```diff
@@ -14,11 +16,28 @@ class DeploymentIndex extends Component {
...
+  }
   render() {
+    var value;
+    if(this.state.isGithub == true) {
+      value = "Not using github?"
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_index.js]: `value` is a pretty generic variable name. Consider a more appropriate name

File: apps/dockup_ui/web/static/js/components/deployment_index.js [hunk 8]
```diff
@@ -14,11 +16,28 @@ class DeploymentIndex extends Component {
     deploymentsArray = deploymentsArray.concat(newDeployment)
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_index.js]: Can `push` work here?

Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_index.js]: It's a good practice to avoid assigning a modified variable to the same variable name. Makes code a little clearer.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 9]
```diff
@@ -19,10 +21,15 @@ class DeploymentForm extends Component {
...
+    if (!this.state.username || !this.state.repository || !this.state.branch) {
+      return "Please enter a github project and branch to deploy";
+    } else if ((this.props.isGithub == false) && (!this.state.url || !this.state.branch)){
+      return "Please enter a project url and branch to deploy";
     } else {
-      return `Click the Deploy button to deploy https://github.com/${this.state.username}/${this.state.repository}`;
+      if ((this.props.isGithub == false) && (this.state.url || this.state.branch)){
+        return `Click the Deploy button to deploy ${this.state.url} with branch: ${this.state.branch}`
+      }
+      return `Click the Deploy button to deploy https://github.com/${this.state.username}/${this.state.repository} with branch: ${this.state.branch}`;
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: These can be extracted out to functions

Comment (iffyuva) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: should there be 2 forms? `easy_github_deploy_formjs` and `generic_git_deploy_form.js` ?

Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Yep, with some care they can be made stateless which would make it easy in reusing code.

File: apps/dockup_ui/web/static/js/components/deployment_index.js (hunk 6 above)
Comment (Atul9) [inline apps/dockup_ui/web/static/js/components/deployment_index.js]: Yes. This can be simplified to this.setState({isGithub: !this.state.isGithub}) It works.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 10]
```diff
@@ -19,29 +26,88 @@ class DeploymentForm extends Component {
...
   displayHelpText() {
-    if (!this.state.username || !this.state.repository) {
-      return "Please enter a github project to deploy";
+    if (!this.state.username || !this.state.repository || !this.state.branch) {
+      return "Please enter a github project and branch to deploy";
+    } else if ((this.props.isGithub == false) && (isAnyFieldEmpty(!this.state.url, !this.state.branch))) {
+      return "Please enter a project url and branch to deploy";
     } else {
-      return `Click the Deploy button to deploy https://github.com/${this.state.username}/${this.state.repository}`;
+      return displayDeployMessage(this.props.isGithub, isAnyFieldEmpty(this.state.url,
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Having a function more than 2 arguments is usually a code smell. Can we pass the props and state directly and let the function figure out what it needs?

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 11]
```diff
@@ -19,29 +26,88 @@ class DeploymentForm extends Component {
   }
 
   displayHelpText() {
-    if (!this.state.username || !this.state.repository) {
-      return "Please enter a github project to deploy";
+    if (!this.state.username || !this.state.repository || !this.state.branch) {
+      return "Please enter a github project and branch to deploy";
+    } else if ((this.props.isGithub == false) && (isAnyFieldEmpty(!this.state.url, !this.state.branch))) {
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: This whole manipulation is still not clear to me. What is the difference between a `github project` and a `project url`?

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 12]
```diff
@@ -19,29 +26,88 @@ class DeploymentForm extends Component {
   }
 
   displayHelpText() {
-    if (!this.state.username || !this.state.repository) {
-      return "Please enter a github project to deploy";
+    if (!this.state.username || !this.state.repository || !this.state.branch) {
+      return "Please enter a github project and branch to deploy";
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: If you're using `return`, the `else` part is unnecessary. The next one can simply be an `if` statement. So something like:


```javascript
if noneFilled(this.state) {
  return "message";
}

if githubNotFilled(this.state, this.props) {
  return "github not filled";
}
```

will do.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 13]
```diff
@@ -19,29 +26,88 @@ class DeploymentForm extends Component {
...
+  }
+
+  handleBranchChange(event) {
+    this.setState({branch: event.target.value});
+  }
+
   render() {
+    let partial;
+    if(this.props.isGithub == true) {
+      partial = <div className="col-md-12">
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: see if this can be extracted out to a [stateless component](https://facebook.github.io/react/docs/reusable-components.html#stateless-functions)

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 14]
```diff
@@ -19,29 +26,88 @@ class DeploymentForm extends Component {
...
+        </b>
+        <input name="branch" class="form-control" onChange={this.handleBranchChange} className="form-control" />
         <div>{this.displayHelpText()}</div>
-        <br/>
+        <button name="deploy" onClick={this.handleClick} disabled={(!this.state.username || !this.state.repository)} className="btn btn-primary">Deploy</button>
       </div>
     )
   }
 }
+let isAnyFieldEmpty = function(field1, field2) {
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: prefer `const` over `lets` for immutable data/functions

File: apps/dockup_ui/web/static/js/components/deployment_form.js (hunk 11 above)
Comment (Atul9) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: The project url is for other source control managemet services like bitbucket.

Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Ah, I see. So if it's a github based deployment, you check for some attributes, and if it's a non-github deployment, you check for others? I really think this form should be broken down into two components. I'll give an example in Ruby:


Say you have one class that handles two different code paths, it might look something like this:


```ruby
class A
  def initialize(input)
    @input = input
  end

  def do_something_conditionally
    if @input
      "Okay, I will do step A"
    else
      "Okay, I will do step B instead"
    end
  end
end
```

This is fine for now. However, imagine if @iffyuva comes up and asks you to add a new feature which needs you to change this class and add a new method called `do_another_thing`. Now, even in that method you need to check for the `@input` and then execute accordingly. Not only this makes the class and methods longer and more complex code-wise, it also makes it hard to understand. Who ever stumbles on this code again (even you, after say 1 month) will have to read through every line and every conditional switch to understand what's going on. This is a pretty bad practice. I'm explaining all this because it's better to unlearn this right now.

A much better pattern would be:

```
class InputPresent
  def initialize(input)
    @input = input
  end

  def do_something
    "Okay, I will do step A"
  end

  def do_another_thing
    "Okay, I will do step A again"
  end
end

class InputNotPresent
  def do_something
    "Okay, I will do step B"
  end

  def do_another_thing
    "Okay, I will do step B again"
  end
end

class A
  def initialize(input)
    @input =
      if input
        InputPresent.new(input)
      else
        InputNotPresent.new
      end
  end

  def do_something
    @input.do_something
  end

  def do_another_thing
    @input.do_another_thing
   end
end
```

Note how we now have only one `if` switch at the top level. The actual methods don't have any checks. Although the line count has increased, it's easier to understand the whole flow.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 15]
```diff
@@ -1,47 +1,82 @@
...
-  handleRepositoryChange(event) {
-    this.setState({repository: event.target.value});
+    this.handleUserNameChange = this.handleUserNameChange.bind(this);
+    this.handleRepositoryChange = this.handleRepositoryChange.bind(this);
+    this.handleClick = this.handleClick.bind(this);
+    this.handleUrlChange = this.handleUrlChange.bind(this);
+    this.handleBranchChange = this.handleBranchChange.bind(this);
   }
 
   displayHelpText() {
```
Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: now that we have separate components, can this function be included in them to avoid that `if this.props.isGithub` check?

Comment (Atul9) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: I have triggered a callback from those components to these methods.

Comment (kgrz) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: But that's the reason why we split the components. Not to have a conditional.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 16]
```diff
@@ -1,47 +1,82 @@
...
 import ReactDOM from "react-dom";
+import GithubComponent from './github_component';
+import GenericComponent from './generic_component';
 
 class DeploymentForm extends Component {
   constructor(props) {
     super(props);
     this.state = {
       username: "",
       repository: "",
```
Comment (emilsoman) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: We should not keep track of username or repository now. That responsibility belongs to the GithubUrlInput component.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 17]
```diff
@@ -1,47 +1,82 @@
 import React, {Component} from 'react';
 import ReactDOM from "react-dom";
+import GithubComponent from './github_component';
```
Comment (emilsoman) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Let's call this "GithubUrlInput" because:

1. We know it's a component since it's in the components directory.
2. GithubComponent means it can be any component related to github. GithubUrlInput makes the intent clearer.

File: apps/dockup_ui/web/static/js/components/deployment_form.js [hunk 18]
```diff
@@ -1,47 +1,82 @@
...
+import GenericComponent from './generic_component';
```
Comment (emilsoman) [inline apps/dockup_ui/web/static/js/components/deployment_form.js]: Same as above. `GenericComponent` is the most *generic* name I've encountered in my programming career 😄  . The person who reads this line has absolutely no damn clue what this component is going to do.

File: apps/dockup_ui/web/static/js/components/github_component.js [hunk 19]
```diff
@@ -0,0 +1,47 @@
+import React, {Component} from 'react';
+
+class GithubComponent extends Component {
```
Comment (emilsoman) [inline apps/dockup_ui/web/static/js/components/github_component.js]: @Atul9 as a learning exercise, let's add a one line comment above all react component classes (or stateless component functions) explaining what the responsibility of the component is - but in less than 80 characters. This helps us strip down responsibilities. Can we give this a try?

File: apps/dockup_ui/web/static/js/components/github_component.js [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+              <td>
+                <b>.git</b>
+              </td>
+            </tr>
+          </tbody>
+        </table>
+      </div>
+    )
+  }
+}
```
Comment (emilsoman) [inline apps/dockup_ui/web/static/js/components/github_component.js]: Please add propTypes for all props passed in to the component. Makes it much easier to understand a component in isolation. Also emits warnings when we pass in wrong props to the component - easier to catch silly mistakes during development this way.

File: apps/dockup_ui/web/static/js/components/github_component.js [hunk 21]
```diff
@@ -0,0 +1,47 @@
+import React, {Component} from 'react';
+
+class GithubComponent extends Component {
+  constructor(props) {
+    super(props);
+    this.handleUserNameChange = this.handleUserNameChange.bind(this);
+    this.handleRepositoryChange = this.handleRepositoryChange.bind(this);
+  }
+
+  handleUserNameChange(event) {
```
Comment (emilsoman) [inline apps/dockup_ui/web/static/js/components/github_component.js]: The parent component is only concerned about the git url. So our GithubUrlInput component should make it its job to construct the git url from the organization and repository and trigger a callback with the git url.

**Comment by emilsoman** [comment]
@Atul9 Please fix the PR comments when you have time

**Comment by emilsoman** [comment]
@Atul9 please resolve conflicts

**Comment by Atul9** [comment]
@emilsoman yes. I will resolve the conflicts today.

**Comment by Atul9** [comment]
@emilsoman I have resolved the conflicts.
